File: aoc24/python/day19.py
from functools import cache

with open("aoc24/inputs/day19.txt") as f:
    patterns, _, *designs = f.read().splitlines()
    patterns = patterns.split(", ")


@cache
def arrangements(design: str):
    if design == "":
        return 1

    return sum(
        arrangements(design.replace(pattern, "", 1))
        for pattern in patterns
        if design.startswith(pattern)
    )


arrs = [arrangements(design) for design in designs]
arrs = [arr for arr in arrs if arr > 0]


print("Part 1:", len(arrs))
print("Part 2:", sum(arrs))

File: aoc24/src/day19.clj
(ns aoc24.src.day19
  (:require
   [aoc-tools :refer [read-input submit-answer]]
   [utils :refer [mfilter]]
   [clojure.string :as str]))

(def input (read-input :test))
(def real-input (read-input))


(defn get-patterns
  [input]
  (->> (str/split (first input) #", ")
       (interleave (repeat 0))
       (partition 2)
       (map reverse)
       (map vec)
       (into {})))

(defn get-designs
  [input]
  (drop 2 input))

(def patterns (get-patterns real-input))
(def designs (get-designs real-input))

(defn possible?
  [design patterns]
  (prn "Checking" design)
  (cond (nil? design) (do (prn "OK") true)
        (patterns design) (do (prn "OK") true)
        :else (let [candidates (->> patterns
                                    (mfilter #(str/starts-with? design %)))
                    _ (prn "Candidatseese for" design "are" candidates)]
                (if (empty? candidates)
                  (do (prn "Not OK") false)
                  (some true? (map #(possible? (str/replace-first design % "") patterns) (sort-by count > (keys candidates))))))))

;; (let [input real-input
;;       designs (get-designs input)
;;       patterns (get-patterns input)]
;;   (->> input
;;        (get-designs)
;;        (take 10)
;;        (map #(possible? % (get-patterns input)))
;;        (filter true?)
;;        (count)
;;        #_(submit-answer 1)))
